code_style.md

```diff
@@ -0,0 +1,125 @@
+- Max line width: 79 chars (with flexibility to overflow by a "few chars" if
+  the overflowing content is not semantically significant and avoids an
+  explosion of vertical whitespace).
+- No trailing whitespace at end of lines.
+- Don't indent empty lines.
+- One newline at the end of the file.
+- Unix newlines, never `\r`
+- Indent similar to our python code: break up long lines at logical boundaries,
+  more than one argument on a line is OK
+- Use semicolons, for consistency with node.
+- UpperCamelCase for class and type names
+- lowerCamelCase for functions and variables.
+- Single line ternary operators are fine.
+- UPPER_CAMEL_CASE for constants
+- Single quotes for strings, for consistency with most JavaScript styles::
+  "bad" // Bad
+  'good' // Good
+- Use parentheses instead of '\\' for line continuation where ever possible
+- Open braces on the same line (consistent with Node):
+  ```
+  if (x) {
+      console.log("I am a fish"); // Good
+  }
+
+  if (x)
+  {
+      console.log("I am a fish"); // Bad
+  }
+  ```
+- Spaces after `if`, `for`, `else` etc, no space around the condition:
+  ```
+  if (x) {
+      console.log("I am a fish"); // Good
+  }
+
+  if(x) {
+      console.log("I am a fish"); // Bad
+  }
+
+  if ( x ) {
+      console.log("I am a fish"); // Bad
+  }
+  ```
+- Declare one variable per var statement (consistent with Node). Unless they
+  are simple and closely related. If you put the next declaration on a new line,
+  treat yourself to another `var`:
+  ```
+  var key = "foo",
+      comparator = function(x, y) {
+          return x - y;
+      }; // Bad
+
+  var key = "foo";
+  var comparator = function(x, y) {
+      return x - y;
+  }; // Good
+
+  var x = 0, y = 0; // Fine
+
+  var x = 0;
+  var y = 0; // Also fine
+  ```
+- A single line `if` is fine, all others have braces. This prevents errors when adding to the code.:
+  ```
+  if (x) return true; // Fine
+
+  if (x) {
+      return true; // Also fine
+  }
+
+  if (x)
+      return true; // Not fine
+  ```
+- Terminate all multi-line lists with commas:
+  ```
+  var mascots = [
+      "Patrick",
+      "Shirley",
+      "Colin",
+      "Susan",
+      "Sir Arthur David" // Bad
+  ];
+
+  var mascots = [
+      "Patrick",
+      "Shirley",
+      "Colin",
+      "Susan",
+      "Sir Arthur David", // Good
+  ];
+  ```
+- Use `null`, `undefined` etc consistently with node:
+  Boolean variables and functions should always be either true or false. Don't set it to 0 unless it's supposed to be a number.
+  When something is intentionally missing or removed, set it to null.
+  Don't set things to undefined. Reserve that value to mean "not yet set to anything."
+  Boolean objects are verboten.
+- Use JSDoc
+
+ECMAScript
+----------
+- Use `let` for variables and `const` for constants. This sounds obvious, but it isn't: the ES6 `const` keyword
+  could be used for assign-once variables, however this guide advises against doing so on the grounds that it
```

Comment:
In Chrome at least, it does (which you could say is the target env for
Vector)

const test = 1;
undefined
test
1
test = 2;
VM3644:1 Uncaught TypeError: Assignment to constant variable.(…)

On Thu, Jul 7, 2016 at 3:09 PM Kegsay <notifications@github.com> wrote:

> In code_style.md
> <https://github.com/matrix-org/matrix-react-sdk/pull/330#discussion_r69878400>
> :
>
> > +      "Colin",
> > +      "Susan",
> > +      "Sir Arthur David", // Good
> > +  ];
> > +  ```
> > +- Use `null`, `undefined` etc consistently with node:
> > +  Boolean variables and functions should always be either true or false. Don't set it to 0 unless it's supposed to be a number.
> > +  When something is intentionally missing or removed, set it to null.
> > +  Don't set things to undefined. Reserve that value to mean "not yet set to anything."
> > +  Boolean objects are verboten.
> > +- Use JSDoc
> > +
> > +ECMAScript
> > +----------
> > +- Use `let` for variables and `const` for constants. This sounds obvious, but it isn't: the ES6 `const` keyword
> > +  could be used for assign-once variables, however this guide advises against doing so on the grounds that it
>
> I do not like using const by default because it is misleading at best,
> and buggy at worst.
>
> It's misleading because the variables that are marked as const *are not
> actually constant*, merely the assignment is (and re-assignment is
> prevented). This is horrendously confusing for people coming from other
> languages, and the amount we as a team context-switch between languages is
> high, so I feel that this is adding needless barriers.
>
> It's buggy because this code *does not throw any errors*:
>
> $ node -v
> v4.2.6
> $ node
> > const foo;
> undefined
> > if (1 === 1) {
> >   foo = 5;
> > }
> > else {
> >   foo = 4;
> > }
> 5
> > foo
> undefined
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/matrix-org/matrix-react-sdk/pull/330/files/56c73b68a9abec549b03232be8dd0e3d2e3f1f9a#r69878400>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAOlRCCfNeY99IfMfIg0KZSJbc4u-Rqwks5qTMlSgaJpZM4JEgSc>
> .
>